[jit] refactoring of module/object

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22504 [WIP] classtypes hold weak_ptr
* #22221 [jit] _script_compile and _script_class_compile add to the python CU
* #22207 [jit] refactor self to be a class again
* #22206 [jit] Give functions qualified names
* #22314 [jit] use save/load for emitFunctionHook
* #22205 [jit] remove unused argument in import.cpp
* #22204 [jit] improvements to QualifiedName
* **#22203 [jit] refactoring of module/object**
* #22202 [jit] Make CompilationUnit own Functions
* #22502 [jit] make test context managers exception safe

Random changes spun out from the big ownership pile:
1. Object should own its compilation unit
2. Some privacy changes and stuff

Differential Revision: [D15998760](https://our.internmc.facebook.com/intern/diff/D15998760)

suo committed:
[jit] refactoring of module/object

Random changes spun out from the big ownership pile:
1. Object should own its compilation unit
2. Some privacy changes and stuff

suo committed:
Update on "[jit] refactoring of module/object"

[jit] refactoring of module/object

Random changes spun out from the big ownership pile:
1. Object should own its compilation unit
2. Some privacy changes and stuff

gh-metadata: pytorch pytorch 22203 gh/suo/67/head

suo committed:
Update on "[jit] refactoring of module/object"

[jit] refactoring of module/object

Random changes spun out from the big ownership pile:
1. Object should own its compilation unit
2. Some privacy changes and stuff

gh-metadata: pytorch pytorch 22203 gh/suo/67/head

suo committed:
Update on "[jit] refactoring of module/object"

[jit] refactoring of module/object

Random changes spun out from the big ownership pile:
1. Object should own its compilation unit
2. Some privacy changes and stuff

gh-metadata: pytorch pytorch 22203 gh/suo/67/head

zdevito commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -268,16 +271,22 @@ struct C10_EXPORT ivalue::Object final : c10::intrusive_ptr_target {
-  Object(std::shared_ptr<ClassType> type, size_t numSlots, OnDelete on_delete)
-      : type_(std::move(type)), on_delete_(on_delete) {
+  Object(
+      std::shared_ptr<torch::jit::script::CompilationUnit> cu,
```
I have vague uncomfort  with the object having so many extra fields and with it not being clear that cu is being kept alive so that the ClassType's methods are kept alive. Can we have a StrongTypePtr similar to StrongFunctionPtr that does the same thing?

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -1052,7 +1052,8 @@ RegisterOperators reg(
            return [type, numAttrs](Stack& stack) {
-             auto userObj = c10::ivalue::Object::create(type, numAttrs);
+             auto userObj =
+                 c10::ivalue::Object::create(type->compilation_unit(), type, numAttrs);
```
`type->compilation_unit()` <- in the future this won't be able to store an owning pointer to the compilation unit. Should we make CompilationUnit into an intrusive_ptr? Otherwise we would need an shared_from_this() here.

suo replied:
I make it a weak_ptr later on, which I think is fine (in particular, we don't want the weak intrusive ptr to accidentally keep the CU alive)

zdevito reviewed (approved):
Looks good.

suo committed:
Update on "[jit] refactoring of module/object"

[jit] refactoring of module/object

Random changes spun out from the big ownership pile:
1. Object should own its compilation unit
2. Some privacy changes and stuff

gh-metadata: pytorch pytorch 22203 gh/suo/67/head

suo committed:
Update on "[jit] refactoring of module/object"

[jit] refactoring of module/object

Random changes spun out from the big ownership pile:
1. Object should own its compilation unit
2. Some privacy changes and stuff

gh-metadata: pytorch pytorch 22203 gh/suo/67/head
